### PR TITLE
Fix _sendMqttMessage implementation to return network error for QoS 0…

### PR DESCRIPTION
**Pull request by aggarw13**
Fix issue causing `IotMqtt_PublishSync`to always return success on network disconnection for when called for **QoS 0**

Related GitHub issue: https://github.com/aws/amazon-freertos/issues/2155

Note: This also fixes the similar behavior of `IotMqtt_SubscribeSync`, and `IotMqtt_UnSubscribeSync` on network disconnection/network send failure

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: libraries/standard/mqtt/src/iot_mqtt_api.c [hunk 1]
```diff
@@ -803,12 +803,22 @@ static IotMqttError_t _sendMqttMessage( _mqttOperation_t * pMqttOperation,
...
+
     IotMqtt_Assert( pMqttOperation != NULL );
 
     /* Send the SUBSCRIBE packet. */
     if( ( flags & MQTT_INTERNAL_FLAG_BLOCK_ON_SEND ) == MQTT_INTERNAL_FLAG_BLOCK_ON_SEND )
     {
         _IotMqtt_ProcessSend( IOT_SYSTEM_TASKPOOL, pMqttOperation->job, pMqttOperation );
+
+        /* Update return status value if operation object status changed after job execution. */
+        if( pMqttOperation->u.operation.status != cachedStatus )
```
Comment (abhidixi11) [inline libraries/standard/mqtt/src/iot_mqtt_api.c:818]: This may be causing double free issues.  Please try moving code to check u.operation.status in PublishAsync() ( for QOS0 and MQTT_INTERNAL_FLAG_BLOCK_ON_SEND )

Comment (aggarw13) [reply libraries/standard/mqtt/src/iot_mqtt_api.c]: Have used a different approach of not utilizing the complex "operations" infrastructure when dealing with the simplistic case of synchronous QoS 0 request.

File: libraries/standard/mqtt/test/unit/iot_tests_mqtt_api.c [hunk 2]
```diff
@@ -282,6 +282,24 @@ static size_t _sendSuccess( IotNetworkConnection_t pSendContext,
...
+static size_t _sendFailure( IotNetworkConnection_t pSendContext,
+                            const uint8_t * pMessage,
+                            size_t messageLength )
+{
+    /* Silence warnings about unused parameters. */
+    ( void ) pSendContext;
+    ( void ) pMessage;
+    ( void ) messageLength;
+
+    /* This function returns the message length to simulate a successful send. */
```
Comment (muneebahmed10) [inline libraries/standard/mqtt/test/unit/iot_tests_mqtt_api.c:297]:
```suggestion
    /* This function returns 0 to simulate a failed send. */
```

File: libraries/standard/mqtt/test/unit/iot_tests_mqtt_api.c [hunk 3]
```diff
@@ -687,12 +705,15 @@ TEST_GROUP_RUNNER( MQTT_Unit_API )
     RUN_TEST_CASE( MQTT_Unit_API, DisconnectAlreadyDisconnected );
     RUN_TEST_CASE( MQTT_Unit_API, PublishQoS0Parameters );
     RUN_TEST_CASE( MQTT_Unit_API, PublishQoS0MallocFail );
+    RUN_TEST_CASE( MQTT_Unit_API, PublishSyncQoS0WithNetworkFailure );
```
Comment (muneebahmed10) [inline libraries/standard/mqtt/test/unit/iot_tests_mqtt_api.c:708]: Nit: `PublishQoS0SyncWithNetworkFailure` seems like it would fit better with the other test names

File: libraries/standard/mqtt/src/iot_mqtt_api.c [hunk 4]
```diff
@@ -1759,6 +1782,101 @@ IotMqttError_t IotMqtt_PublishAsync( IotMqttConnection_t mqttConnection,
...
+    else if( _IotMqtt_ValidatePublish( mqttConnection->awsIotMqttMode,
+                                       pPublishInfo,
+                                       flags,
+                                       pCallbackInfo,
+                                       pPublishOperation ) == false )
+    {
+        status = IOT_MQTT_BAD_PARAMETER;
+    }
+    else if( ( ( flags & MQTT_INTERNAL_FLAG_BLOCK_ON_SEND )
+               == MQTT_INTERNAL_FLAG_BLOCK_ON_SEND ) &&
```
Comment (leegeth) [inline libraries/standard/mqtt/src/iot_mqtt_api.c:1861]: Are we expecting that the flag MQTT_INTERNAL_FLAG_BLOCK_ON_SEND  is set from application? Isn't this an internal flag and I don't see it in the public docs as well https://docs.aws.amazon.com/freertos/latest/lib-ref/c-sdk/mqtt/mqtt_constants.html#mqtt_constants_flags

File: libraries/standard/mqtt/src/iot_mqtt_api.c [hunk 5]
```diff
@@ -280,6 +280,35 @@ static void _handleConnectFailure( IotMqttConnection_t pMqttConnection,
...
+ * @param[in] mqttConnection The MQTT connection to use for the PUBLISH.
+ * @param[in] pPublishInfo MQTT publish parameters.
+ * @param[in] flags Flags which modify the behavior of this function.
+ * @param[in] pCallbackInfo Asynchronous notification of this function's completion (`NULL` to disable).
+ * @param[out] pPublishOperation Set to a handle by which this operation may be
+ * referenced after this function returns. This reference is invalidated once
+ * the publish operation completes.
+ *
+ * @return This function will return #IOT_MQTT_STATUS_PENDING upon success for
+ * QoS 1 publishes
```
Comment (leegeth) [inline libraries/standard/mqtt/src/iot_mqtt_api.c:294]: How about the return codes when there are errors?

File: libraries/standard/mqtt/src/iot_mqtt_api.c [hunk 6]
```diff
@@ -280,6 +280,35 @@ static void _handleConnectFailure( IotMqttConnection_t pMqttConnection,
...
+                                     uint32_t flags,
+                                     const IotMqttCallbackInfo_t * pCallbackInfo,
+                                     IotMqttOperation_t * const pPublishOperation );
+
+/**
+ * @brief Sends a Publish message at QoS 0 synchronously.
+ * @param[in] mqttConnection The MQTT connection to use for the PUBLISH.
+ * @param[in] pPublishInfo MQTT publish parameters.
+ *
+ * @return returns #IOT_MQTT_SUCCESS upon success.
```
Comment (leegeth) [inline libraries/standard/mqtt/src/iot_mqtt_api.c:307]: Same comment about the return values for error cases.

File: libraries/standard/mqtt/src/iot_mqtt_api.c [hunk 7]
```diff
@@ -1750,6 +1763,16 @@ IotMqttError_t IotMqtt_PublishAsync( IotMqttConnection_t mqttConnection,
...
         }
 
+        if( ( pPublishInfo->qos == IOT_MQTT_QOS_0 ) && ( pOperation != NULL ) )
+        {
+            if( ( ( flags & MQTT_INTERNAL_FLAG_BLOCK_ON_SEND )
+                  == MQTT_INTERNAL_FLAG_BLOCK_ON_SEND ) &&
+                ( pOperation->u.operation.status == IOT_MQTT_NETWORK_ERROR ) )
+            {
+                status = IOT_MQTT_NETWORK_ERROR;
+            }
```
Comment (leegeth) [inline libraries/standard/mqtt/src/iot_mqtt_api.c:1773]: In what case, the execution would reach here? Isn't it always be going to _publishQos0Sync as per line 1864.

File: libraries/standard/mqtt/src/iot_mqtt_api.c [hunk 8]
```diff
@@ -1652,27 +1681,11 @@ IotMqttError_t IotMqtt_PublishAsync( IotMqttConnection_t mqttConnection,
...
     uint8_t ** pPacketIdentifierHigh = NULL;
 
-    /* Check that IotMqtt_Init was called. */
-    if( _checkInit() == false )
-    {
-        status = IOT_MQTT_NOT_INITIALIZED;
-    }
-    else if( _IotMqtt_ValidatePublish( mqttConnection->awsIotMqttMode,
-                                       pPublishInfo,
+    /* Create a PUBLISH operation. */
```
Comment (leegeth) [inline libraries/standard/mqtt/src/iot_mqtt_api.c:1684]: Better to have asserts in the internal function to validate the arguments.

File: libraries/standard/mqtt/src/iot_mqtt_api.c [hunk 9]
```diff
@@ -1759,6 +1782,101 @@ IotMqttError_t IotMqtt_PublishAsync( IotMqttConnection_t mqttConnection,
...
+                                       pPublishOperation ) == false )
+    {
+        status = IOT_MQTT_BAD_PARAMETER;
+    }
+    else if( ( ( flags & MQTT_INTERNAL_FLAG_BLOCK_ON_SEND )
+               == MQTT_INTERNAL_FLAG_BLOCK_ON_SEND ) &&
+             ( pPublishInfo->qos == IOT_MQTT_QOS_0 ) )
+    {
+        status = _publishQos0Sync( mqttConnection,
+                                   pPublishInfo );
```
Comment (leegeth) [inline libraries/standard/mqtt/src/iot_mqtt_api.c:1865]: What is the callback behavior in this case? Are we completely ignoring the callback behavior? What will happen to applications already present and expecting a callback to mark the completion of publishAsync?

File: libraries/standard/mqtt/src/iot_mqtt_api.c (hunk 4 above)
Comment (aggarw13) [reply libraries/standard/mqtt/src/iot_mqtt_api.c:1861]: Good catch. Will remove this logic, was added in the initial revision of the fix in the PR, but had forgotten to remove it.

Regarding your point that the flag is internal facing, unfortunately, the `IotMqtt_PublishAsync` API will accept the flag even if the user accidentally sets the bit associated with this flag in the `flags` input parameter. (This is the behavior of `v4_beta` branch)

File: libraries/standard/mqtt/src/iot_mqtt_api.c (hunk 7 above)
Comment (aggarw13) [reply libraries/standard/mqtt/src/iot_mqtt_api.c:1773]: Yes, will remove this logic as mentioned in https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1009#discussion_r445779256

File: libraries/standard/mqtt/test/unit/iot_tests_mqtt_api.c (hunk 2 above)
Comment (aggarw13) [reply libraries/standard/mqtt/test/unit/iot_tests_mqtt_api.c]: Changed

File: libraries/standard/mqtt/test/unit/iot_tests_mqtt_api.c (hunk 3 above)
Comment (aggarw13) [reply libraries/standard/mqtt/test/unit/iot_tests_mqtt_api.c]: Changed

File: libraries/standard/mqtt/src/iot_mqtt_api.c (hunk 5 above)
Comment (aggarw13) [reply libraries/standard/mqtt/src/iot_mqtt_api.c]: Added documentation.

File: libraries/standard/mqtt/src/iot_mqtt_api.c (hunk 8 above)
Comment (aggarw13) [reply libraries/standard/mqtt/src/iot_mqtt_api.c:1703]: Added assert checks.

File: libraries/standard/mqtt/src/iot_mqtt_api.c (hunk 4 above)
Comment (leegeth) [reply libraries/standard/mqtt/src/iot_mqtt_api.c:1870]: I see that it is a possible behavior for the async API. Isn't this flag supposed to be set internally though for a publish sync API call for QoS0? Are you missing that set logic in this PR?

File: libraries/standard/mqtt/src/iot_mqtt_api.c (hunk 6 above)
Comment (aggarw13) [reply libraries/standard/mqtt/src/iot_mqtt_api.c]: Added

File: libraries/standard/mqtt/src/iot_mqtt_api.c (hunk 4 above)
Comment (leegeth) [reply libraries/standard/mqtt/src/iot_mqtt_api.c:1868]: As discussed, I see that the flag is set from publish sync API

File: libraries/standard/mqtt/src/iot_mqtt_api.c (hunk 9 above)
Comment (leegeth) [reply libraries/standard/mqtt/src/iot_mqtt_api.c:1875]: As discussed offline, as this path is hit only for sync API, callbacks are not going to be present.

Comment (aggarw13) [reply libraries/standard/mqtt/src/iot_mqtt_api.c:1875]: We talked separately to clarify that a callback parameter is not expected for the sync Publish case as firstly, the `MQTT_INTERNAL_FLAG_BLOCK_ON_SEND` flag is not public-facing and thus, the customer cannot call the `IotMqtt_PublishAsync` function for a QoS 0 synchronous operation, and secondly, the sync Publish Qos 0 operation is requested by the `IotMqtt_PublishSync` function that passes the internal `MQTT_INTERNAL_FLAG_BLOCK_ON_SEND` flag to the `IotMqtt_PublishAsync` function.

Basically, the changes made do not change the behavior of the function (for any customer application depending on this function to be affected)!
